### PR TITLE
feat(stepfunctions): adding custom state name prop

### PR DESCRIPTION
**Pull request by SankyRed**
Closes https://github.com/aws/aws-cdk/issues/23532

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/aws-cdk-lib/aws-stepfunctions/lib/states/fail.ts [hunk 1]
```diff
@@ -9,6 +9,8 @@ import { INextable } from '../types';
 export interface FailProps {
   /**
    * Optional name for this state
+   *
+   * @default LogicalResourceId
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/lib/states/fail.ts:13]: I don't think that this is accurate. It's more accurate to say something like:

```ts
* @default - The construct ID will be used as state name
```

File: packages/aws-cdk-lib/aws-stepfunctions/lib/states/succeed.ts [hunk 2]
```diff
@@ -54,6 +61,7 @@ export class Succeed extends State {
   public toStateJson(): object {
     return {
       Type: StateType.SUCCEED,
+      StateName: this.stateName,
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/lib/states/succeed.ts:64]: If we put `StateName` in this return object, what ultimately happens to it?

Have you seen the documentation for [Amazon States Language](https://docs.aws.amazon.com/step-functions/latest/dg/concepts-amazon-states-language.html)?

This is what we are targeting. It looks like this:

```ts
{
  "StartAt": "<StateName>",
  "States": {
    "<StateName1>": {
      "Type": "<Whatever>",
      <Properties>
    },
    "<StateName2>": {
      "Type": "<MoreWhatever>",
    }
}
```

That is ultimately what we want to change.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

File: packages/aws-cdk-lib/aws-stepfunctions/lib/states/fail.ts (hunk 1 above)
Comment (SankyRed) [reply packages/aws-cdk-lib/aws-stepfunctions/lib/states/fail.ts:13]: Okay, will update that. Thanks Rico.

**Review by kaizencc — COMMENTED** [review]
Can you try again with the title? Couple things:

- this is a feature, not a chore. if you feel like you need exemptions for integ tests, you can add the relevant labels yourself.
- standard is to not capitalize words
- since you are working in a particular module, the standard is to include the module name in the title (i.e. `feat(stepfunctions):`)
- try to rework the title to be more descriptive and describe the feature that has been added. if you've followed the other steps we're now at "feat(stepfunctions): adding custom state name prop". I don't think people who read the title will immediately know where the custom state name prop can be added.

File: packages/aws-cdk-lib/aws-stepfunctions/README.md [hunk 3]
```diff
@@ -124,6 +124,7 @@ becomes new the new Data object. This behavior can be modified by supplying valu
 
 These properties impact how each individual step interacts with the state machine data:
 
+* `stateName`: the custom state name to each state if required else we used the construct id
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/README.md:127]: "if required"

Required by who?

Comment (rix0rrr) [reply packages/aws-cdk-lib/aws-stepfunctions/README.md:127]: Please use punctuation as well.

"the name of the state in the state machine definition. If not supplied, defaults to the construct id."

Or something.

File: packages/aws-cdk-lib/aws-stepfunctions/README.md [hunk 4]
```diff
@@ -249,6 +250,8 @@ An arbitrary JSON object (specified at execution start) is passed from state to
 state and transformed during the execution of the workflow. For more
 information, see the States Language spec.
 
+You can add custom state names to all states and if not it defaults to construct id.
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/README.md:253]: Did we not already say this?

File: packages/aws-cdk-lib/aws-stepfunctions/README.md [hunk 5]
```diff
@@ -281,10 +284,12 @@ The `Pass` state also supports passing key-value pairs as input. Values can
 be static, or selected from the input with a path.
 
 The following example filters the `greeting` field from the state input
-and also injects a field called `otherData`.
+and also injects a field called `otherData`. You can also add custom state name
+to Pass state.
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/README.md:288]: I don't think we need to add this everywhere.

File: packages/aws-cdk-lib/aws-stepfunctions/README.md [hunk 6]
```diff
@@ -281,10 +284,12 @@ The `Pass` state also supports passing key-value pairs as input. Values can
...
 
 ```ts
 const pass = new sfn.Pass(this, 'Filter input and inject data', {
+  stateName: 'my-pass-state',
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/README.md:292]: Leave this out I think? Or give a single example on `stateName`, adding a comment saying what the default is.

File: packages/aws-cdk-lib/aws-stepfunctions/lib/states/choice.ts [hunk 7]
```diff
@@ -47,7 +54,7 @@ export class Choice extends State {
   public readonly endStates: INextable[] = [];
 
   constructor(scope: Construct, id: string, props: ChoiceProps = {}) {
-    super(scope, id, props);
+    super(scope, props.stateName? props.stateName: id, props);
```
Comment (rix0rrr) [inline packages/aws-cdk-lib/aws-stepfunctions/lib/states/choice.ts:57]: This changes the construct ID though.

So passing `stateName` now has 2 effects, affecting the construct ID as well as the state name.

It's just an override for `id` now, which is not what we want.

**Review by rix0rrr — CHANGES_REQUESTED** [review]

**Review by rix0rrr — APPROVED** [review]
Excellent!

**Pull request merged**